### PR TITLE
Use array-based tag table to improve search performance

### PR DESCRIPTION
**Pull request by charmander**
This might make searches take less than 20 seconds. If it does, we should be able to use it for tag blacklists, too.

### CODE REVIEW

File: weasyl/search.py [hunk 1]
```diff
@@ -99,6 +90,15 @@ def add_nonempty(s, item):
             tag = d.get_search_tag(criterion)
             add_nonempty(self.required_includes, tag)
 
+    def __bool__(self):
```
Comment (weykent) [inline weasyl/search.py]: Sadly, python 2 and 3 differ here. This needs to be available as both `__bool__` and `__nonzero__`.

File: weasyl/search.py [hunk 2]
```diff
@@ -269,18 +280,33 @@ def make_statement(statement_select, statement_additional_where, statement_order
...
+        "userid": userid,
+        "rating": rating,
+        "ratings": list(search.ratings),
+        "category": cat,
+        "subcategory": subcat,
+        "limit": limit,
+        "backid": backid,
+        "nextid": nextid,
+        "required_include_count": len(search.required_includes),
+    }
```
Comment (weykent) [inline weasyl/search.py]: Gosh this whole thing makes me sad. Would it be easier to move to query-builder style now? Not that it has to happen in this PR.

Comment (charmander) [inline weasyl/search.py]: Yes. It’d also be nice to use the same logic for all tables. I’ll tackle both soon =)

File: weasyl/searchtag.py [hunk 3]
```diff
@@ -197,6 +208,17 @@ def associate(userid, tags, submitid=None, charid=None, journalid=None):
                 [table, targetid, d.sql_number_list(list(added))])
 
     if submitid:
+        try:
```
Comment (strain-113) [inline weasyl/searchtag.py]: Add a TODO here to replace with an upsert when we're on 9.5?

**Pull request merged**